File: src/part1/Rational.java
package part1;

import static java.lang.Math.abs;

@SuppressWarnings("WeakerAccess")
public final class Rational {
    private final int numerator;

    private final int denominator;

    static private int gcd(int a, int b) {
        if (a == b || b == 0) return a;
        else if (a == 0) return b;
        else if (a > b) return gcd(a % b, b);
        else return gcd(a, b % a);
    }

    public Rational(int numerator, int denominator) {
        if (denominator == 0) throw new ArithmeticException("Denominator cannot be zero");
        int gcd = gcd(abs(numerator), abs(denominator));
        if (denominator < 0) gcd = -gcd;
        this.numerator = numerator / gcd;
        this.denominator = denominator / gcd;
    }

    public int getNumerator() {
        return numerator;
    }

    public int getDenominator() {
        return denominator;
    }

    public Rational plus(Rational other) {
        return new Rational(
                numerator * other.denominator + denominator * other.numerator,
                denominator * other.denominator
        );
    }

    public Rational unaryMinus() {
        return new Rational(-numerator, denominator);
    }

    public Rational minus(Rational other) {
        return plus(other.unaryMinus());
    }

    public Rational times(Rational other) {
        return new Rational(numerator * other.numerator, denominator * other.denominator);
    }

    public Rational div(Rational other) {
        return new Rational(numerator * other.denominator, denominator * other.numerator);
    }

    public int toInt() {
        return numerator / denominator;
    }

    public double toDouble() {
        return ((double) numerator ) / denominator;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj instanceof Rational) {
            Rational other = (Rational) obj;
            return numerator == other.numerator && denominator == other.denominator;
        }
        return false;
    }

    @Override
    public int hashCode() {
        int result = numerator;
        result = 31 * result + denominator;
        return result;
    }

    @Override
    public String toString() {
        return "" + numerator + "/" + denominator;
    }
}

File: src/part1/kotlin/Rational.kt
package part1.kotlin

import java.lang.Math.abs

class Rational(numerator: Int, denominator: Int) {
    private val numerator: Int

    private val denominator: Int

    private tailrec fun gcd(a: Int, b: Int): Int =
            when {
                a == b || b == 0 -> a
                a == 0 -> b
                a > b -> gcd(a % b, b)
                else -> gcd(a, b % a)
            }

    init {
        if (denominator == 0) throw ArithmeticException("Denominator cannot be zero")
        var gcd = gcd(abs(numerator), abs(denominator))
        if (denominator < 0) gcd = -gcd
        this.numerator = numerator / gcd
        this.denominator = denominator / gcd
    }

    operator fun plus(other: Rational) = Rational(
            numerator * other.denominator + denominator * other.numerator,
            denominator * other.denominator
    )

    operator fun unaryMinus() = Rational(-numerator, denominator)

    operator fun minus(other: Rational) = plus(other.unaryMinus())

    operator fun times(other: Rational) = Rational(numerator * other.numerator, denominator * other.denominator)

    operator fun div(other: Rational) = Rational(numerator * other.denominator, denominator * other.numerator)

    fun toInt() = numerator / denominator

    fun toDouble() = numerator.toDouble() / denominator

    override fun equals(other: Any?) =
            when {
                this === other -> true
                other is Rational -> numerator == other.numerator && denominator == other.denominator
                else -> false
            }

    override fun hashCode(): Int {
        var result = numerator
        result = 31 * result + denominator
        return result
    }

    override fun toString() = "$numerator/$denominator"
}